lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -0,0 +1,416 @@
+
+namespace llvm {
+template <>
+struct DenseMapInfo<Node> {
+  static Node getEmptyKey() {
+    auto *pointer = llvm::DenseMapInfo<void *>::getEmptyKey();
+    return Node(Value::getFromOpaquePointer(pointer), 0);
+  }
+  static Node getTombstoneKey() {
+    auto *pointer = llvm::DenseMapInfo<void *>::getTombstoneKey();
+    return Node(Value::getFromOpaquePointer(pointer), 0);
+  }
+  static unsigned getHashValue(const Node &node) {
+    return detail::combineHashValue(mlir::hash_value(node.val),
+                                    llvm::hash_value(node.fieldId));
+  }
+  static bool isEqual(const Node &lhs, const Node &rhs) { return lhs == rhs; }
+};
+} // namespace llvm
+
+class ModuleDFS {
+public:
+  ModuleDFS(FModuleOp module, InstanceGraph &instanceGraph,
+            llvm::SmallDenseMap<Node, SmallVector<Node>> &portPaths)
+      : module(module), instanceGraph(instanceGraph), portPaths(portPaths) {}
+
+  LogicalResult dfsFromNode(Node &node, Node inputArg) {
+
+    LLVM_DEBUG(llvm::dbgs()
+               << "\n visiting node :" << node.val << "," << node.fieldId);
+    if (!node.val.getType().isa<FIRRTLBaseType>())
+      return success();
+    if (visited.contains(node))
+      return success();
+    // visiting: Set of nodes that have a path to this current node.
+    if (visiting.contains(node)) {
+      FieldRef f(node.val, node.fieldId);
+      auto signalName = getFieldName(f);
+      auto errorDiag = mlir::emitError(
+          module.getLoc(),
+          "detected combinational cycle in a FIRRTL module, sample path: ");
+      errorDiag.attachNote(node.val.getLoc());
+      if (!signalName.empty())
+        errorDiag << module.getName() << "." << signalName;
+      for (auto n : llvm::reverse(currentPath)) {
+        FieldRef f(n.val, n.fieldId);
+        auto signalName = getFieldName(f);
+        if (!signalName.empty())
+          errorDiag << " -> " << module.getName() << "." << signalName;
+        if (n == node)
+          break;
+      }
+      return failure();
+    }
+    visiting.insert(node);
+    currentPath.push_back(node);
+    // Case 1: Instance and Mem ops.
+    if (!handleDefOps(node, inputArg)) {
+      // Case 2: Handle aggregate values.
+      if (!node.val.getType().cast<FIRRTLBaseType>().isGround()) {
+        auto baseVal = node.val;
+        size_t fieldId = node.fieldId;
+        if (node.fieldId == 0) {
+          // Get the base aggregate value from the leaf.
+          auto fRef = getFieldRefFromValue(node.val);
+          baseVal = fRef.getValue();
+          fieldId = fRef.getFieldID();
+        }
+        if (!valLeafOps.count(baseVal))
+          gatherAggregateLeafs(baseVal);
+        // Get all the leaf values, that correspond to the `fieldId` leaf of the
+        // `baseVal`. This is required, because, mulitple subfield ops can refer
+        // to the same field.
+        for (auto &leaf : valLeafOps[baseVal][fieldId]) {
```

Comment:
Iterate over DenseMap would cause non-deterministic behaivor. Can we use MapVector or something?